src/ParallelAlgorithms/EventsAndTriggers/Actions/RunEventsOnFailure.hpp

```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <optional>
+#include <tuple>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Tags.hpp"
+#include "Utilities/CloneUniquePtrs.hpp"
+#include "Utilities/ErrorHandling/FloatingPointExceptions.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Actions {
+/*!
+ * \brief Invokes all events specified in `Tags::EventsRunAtCleanup`.
+ *
+ * Before running the events, floating point exceptions are disabled. This is
+ * to allow manipulating data even if there are `NaN` or other problematic
+ * values. We ultimately just want to be able to see the state of the
+ * simulation at failure.
+ *
+ * This action is intended to be in the `Parallel::Phase::PostFailureCleanup`
+ * phase.
+ *
+ * \note When using local time stepping, the simulation will almost certainly
+ * fail with different elements have different times. This can also happen
+ * with global time stepping if some elements updated their current time
+ * before other elements hit an error. If an event requires consistent
+ * observation IDs, such as ObserveFields, an option to override the received
```

Comment:
then an option should be added to that event that specifies what should be used instead of the received observation IDs.  (See the ... option of ObserveFields as an example.)